primitives/wasm-interface/src/lib.rs

```diff
@@ -338,10 +357,25 @@ pub trait Sandbox {
 	fn get_global_val(&self, instance_idx: u32, name: &str) -> Result<Option<Value>>;
 }
 
+if_wasmtime_is_enabled! {
+	pub trait HostFunctionRegistry {
+		type State;
+		type Error;
+		type FunctionContext: FunctionContext;
+		fn with_function_context<R>(caller: wasmtime::Caller<Self::State>, callback: impl FnOnce(&mut dyn FunctionContext) -> R) -> R;
```

Comment:
Making it `Self::FunctionContext` significantly complicates the lifetimes everywhere, and from the top of my head I'm not even sure if it's actually possible to use a concrete type here without GATs (since `Caller` actually takes a lifetime which has to be passed to the `Self::FunctionContext`, it's just elided here). Also, rest of the traits (e.g. `host::FromFFIValue`) and the dynamic dispatch machinery (`Function::execute`) take it as `&dyn FunctionContext` anyway.

If we really want a concrete type here I'd suggest that we should ignore it for now and just revisit this once GATs are stabilized.

Reply:
Nah, all is fine. Was just curious